File: Crio/crio_programming_interview_problems-master/SumLists2/Solution.cpp
/*
class ListNode {
    public:
	int val;
	ListNode* next;

	ListNode() {}

	ListNode(int x) {                      
	    val = x;                       
	    next = NULL;
	}
};
*/
#include "../crio/cpp/ds/ListNode/ListNode.hpp"

// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
ListNode* reverseLinkedList(ListNode* node) {
    ListNode* rev = NULL;
    while(node != NULL) {
        ListNode* temp = node;
        node = node->next;
        temp->next = rev;
        rev = temp;
    }
    return rev;
}
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

ListNode* sumLists2(ListNode* head1 , ListNode* head2) {
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    int carry = 0;
    head1 = reverseLinkedList(head1);
    head2 = reverseLinkedList(head2);
    ListNode* result = new ListNode(0);
    ListNode* current = result;
    ListNode* prev = current;
    while(head1 != NULL && head2 != NULL) {
        current->val = (head1->val + head2->val + carry) % 10;
        carry = (head1->val + head2->val + carry) / 10;
        current->next = new ListNode(0);
        prev = current;
        current = current->next;
        head1 = head1->next;
        head2 = head2->next;
    }

    while(head1 != NULL) {
        current->val = (head1->val + carry) % 10;
        carry = (head1->val + carry) / 10;
        current->next = new ListNode(0);
        prev = current;
        current = current->next;
        head1 = head1->next;
    }

    while(head2 != NULL) {
        current->val = (head2->val + carry) % 10;
        carry = (head2->val + carry) / 10;
        current->next = new ListNode(0);
        prev = current;
        current = current->next;
        head2 = head2->next;
    }

    if(carry > 0) {
        current->val = carry;
    } else {
        prev->next = NULL;
    }
    return reverseLinkedList(result);

// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}

File: Crio/crio_programming_interview_problems-master/SumLists2/Solution.java
/*
public class ListNode {
   public int val; 
   public ListNode next;
   public ListNode(int x) { val = x; next = null; }
}
*/
import java.io.*;
import java.util.*;
import crio.ds.List.ListNode;
class Solution {
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
  public static ListNode reverseList( ListNode head) {
    ListNode prev = null , nxt;

    while (head != null) {
      nxt = head.next;
      head.next = prev;
      prev = head;
      head = nxt;
    }
    return prev;
  }
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
  public static ListNode sumLists1( ListNode head1 , ListNode head2) {
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    head1 = reverseList(head1);
    head2 = reverseList(head2);
    ListNode ans = new ListNode(0);
    ListNode ansHead = ans;

    int sizeOfFirst = 0 , sizeOfSecond = 0;
    ListNode temp = head1;
    while (temp != null) {
      sizeOfFirst++;
      temp = temp.next;
    }

    temp = head2;
    while (temp != null) {
      sizeOfSecond++;
      temp = temp.next;
    }

    boolean flag = false;
    int carry = 0;
    int minSize = Math.min(sizeOfFirst, sizeOfSecond);
    for ( int i = 0; i < minSize; ++i) {
      int x = head1.val + head2.val;
      if ( flag == false) {
        ans.val = x % 10;
        carry = x / 10; flag = true;
        head1 = head1.next;
        head2 = head2.next;
        continue;
      }
      int rem = (carry + x % 10) % 10;
      ans.next = new ListNode(rem);
      ans = ans.next;
      carry = (carry + x ) / 10;
      head1 = head1.next;
      head2 = head2.next;
    }
    while (head1 != null) {
      int x = head1.val;
      int rem = (carry + x % 10) % 10;
      carry = (carry + x  ) / 10;
      ans.next = new ListNode(rem);
      ans = ans.next;
      head1 = head1.next;
    }

    while (head2 != null) {
      int x = head2.val;
      int rem = (carry + x % 10) % 10;
      carry = (carry + x  ) / 10;
      ans.next = new ListNode(rem);
      ans = ans.next;
      head2 = head2.next;
    }
    if ( carry > 0) {
      ans.next = new ListNode(carry);
    }
    ansHead = reverseList(ansHead);
    temp = ansHead;
    return ansHead;

// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
  }
}
